salt/modules/virt.py

```diff
@@ -599,18 +601,29 @@ def _gen_xml(
                 context["boot"]["kernel"] = "/usr/lib/grub2/x86_64-xen/grub.xen"
                 context["boot_dev"] = []
 
-    if "serial_type" in kwargs:
-        context["serial_type"] = kwargs["serial_type"]
-    if "serial_type" in context and context["serial_type"] == "tcp":
-        if "telnet_port" in kwargs:
-            context["telnet_port"] = kwargs["telnet_port"]
-        else:
-            context["telnet_port"] = 23023  # FIXME: use random unused port
-    if "serial_type" in context:
-        if "console" in kwargs:
-            context["console"] = kwargs["console"]
-        else:
-            context["console"] = True
+    context["serials"] = []
+    for serial in enumerate(serials):
+        serial_context = {
+            "type": serial.type
+        }
+        if serial_context["type"] == "tcp":
+            if (serial["port"] is None) or (serial["protocol"] is None):
+                raise CommandExecutionError("Absent serial parameter for TCP connection")
+            serial_context["port"] = serial["port"]
+            serial_context["protocol"] = serial["protocol"]
+        context["serials"].append(serial_context)
+
+    context["consoles"] = []
+    for console in enumerate(consoles):
+        console_context = {
+            "type": console.type
+        }
+        if console_context["type"] == "tcp":
+            if (console["port"] is None) or (console["protocol"] is None):
+                raise CommandExecutionError("Absent console parameters for TCP connection")
+            console_context["port"] = console["port"]
+            console_context["protocol"] = console["protocol"]
```

Comment:
In the old code the protocol is forced to `telnet`. The port was also set to the same default that the serial port (`23023`). You need to use it as a default values.
```suggestion
            console_context["port"] = console.get("port", default_port)
            console_context["protocol"] = console.get("protocol", "telnet")
```